jupyterhub/app.py

```diff
@@ -2358,6 +2359,195 @@ async def init_api_tokens(self):
+            host (Optional[str]): The host of Hub session. Defaults to None.
+
+        Returns:
+            Optional[Service]: The created service
+        """
+
+        if domain is None:
+            if self.domain:
+                domain = 'services.' + self.domain
+            else:
+                domain = ''
+
+        if host is None:
+            if self.domain:
+                parsed = urlparse(self.subdomain_host)
+                host = f'{parsed.scheme}://services.{parsed.netloc}'
+            else:
+                host = ''
+
+        if 'name' not in spec:
+            raise ValueError('service spec must have a name: %r' % spec)
+
+        name = spec['name']
+        # get/create orm
+        orm_service = orm.Service.find(self.db, name=name)
+        if orm_service is None:
+            # not found, create a new one
+            orm_service = orm.Service(name=name, from_config=from_config)
+            self.db.add(orm_service)
+            if spec.get('admin', False):
+                self.log.warning(
+                    f"Service {name} sets `admin: True`, which is deprecated in JupyterHub 2.0."
+                    " You can assign now assign roles via `JupyterHub.load_roles` configuration."
+                    " If you specify services in the admin role configuration, "
+                    "the Service admin flag will be ignored."
+                )
+                roles.update_roles(self.db, entity=orm_service, roles=['admin'])
+        else:
+            # Do nothing if the config file tries to modify a API-base service
+            # or vice versa.
+            if orm_service.from_config != from_config:
+                return
+        orm_service.admin = spec.get('admin', False)
+
+        self.db.commit()
+        service = Service(
+            parent=self,
+            app=self,
+            base_url=self.base_url,
+            db=self.db,
+            orm=orm_service,
+            roles=orm_service.roles,
+            domain=domain,
+            host=host,
+            hub=self.hub,
+        )
+
+        traits = service.traits(input=True)
+        for key, value in spec.items():
+            if key not in traits:
+                raise AttributeError("No such service field: %s" % key)
+            setattr(service, key, value)
+            orm_service.update_column(key, value)
+
+        if service.api_token:
+            self.service_tokens[service.api_token] = service.name
+        elif service.managed:
+            # generate new token
+            # TODO: revoke old tokens?
+            service.api_token = service.orm.new_api_token(note="generated at startup")
+
+        if service.url:
+            parsed = urlparse(service.url)
+            if parsed.port is not None:
+                port = parsed.port
+            elif parsed.scheme == 'http':
+                port = 80
+            elif parsed.scheme == 'https':
+                port = 443
```

Comment:
else: raise an Exception?

Reply:
> I want to avoid raising an exception in this function to keep Jupyterhub from crashing at startup because of badly configured services.

I don't think we should do that. If config is invalid and not loaded, it's a lot more surprising to see deployment and everything succeed and your config just have no effect than it is to see an error.

Given how JupyterHub is usually deployed, logs don't tend to get seen until _after_ a problem is noticed by real users. So admin errors like invalid service config should probably prevent startup, because otherwise bad config is more likely to go unnoticed long enough to affect _users_.